README.md
# can-cube-timer
Simple Rubik's Cube timer


# TODO
* More events for timer status change and solve phase change
* Think more on the constructor api. should it be exposing all of that functionality
* should there be a constructor overload without the opts object?
* Should it return a promise when you call start that will be resolved when the solve ends?
this would be useful but also feels like doing too many things


```javascript
import canCubeTimer, { SOLVING, INSPECTING } from 'can-cube-timer';

const inSeconds = ms => ms / 1000;

// See the full documentation for other options
const timer = canCubeTimer({
  onInspectionWarning: timeRemaining => console.log(`You have ${inSeconds(timeRemaining)} seconds remaining!`),
  onPenalty: (isDNF, timeRemaining) => isDNF ? 
    console.log('Maximum allowed time elapsed. Marking as DNF') :
    console.log(`15 seconds past, +2 second penalty, ${timeRemaining} seconds remaining until inspection end`,
  onTick: (phase, time, timeRemaining) => {
    console.log(`Currently ${phase === SOLVING : 'solv' ? 'inspect'}ing!`);
    console.log(`Current time is ${inSeconds(time)}`);
    
    // timeRemaining will be null if phase is SOLVING
    console.log(`Inspection time remaining ${timeRemaining}`);
  },
  onDone: ({ inspectionTime, isDNF, penalized, solveTime }) => {
    // see api documentation on solveTime
    console.log(```
      Inspection Time: ${inSeconds(inspectionTime)}
      DNF: ${isDNF}
      Penalized: ${isPenalized}
      Solve time: ${inSeconds(solveTime)}
    ```);
  }
});


timer.start();

// ...
const inspectionPeek = timer.peek();
console.log(`Timer status: ${inspectionPeek.status}`);
console.log(`Timer phase: ${inspectionPeek.phase}`);
console.log(`Current time: ${inspectionPeek.time}`); // This is the current time of the current phase.`);
console.log(`Total time: ${inspectionTime.totalTime}`); // time since the .start() call

const inspectionStats = timer.startSolve(); // See the documentation on .start() and .startSolve();
// TODO: console.logs for inspectionStats
// ...

const solvePeek = timer.status();

console.log(`Timer status: ${solvePeek.status}`);
console.log(`Timer phase: ${solvePeek.phase}`);
console.log(`Current time: ${solvePeek.time}`); // This is the current time of the current phase.`);
console.log(`Total time: ${solvePeek.totalTime}`); // time since the .start() call


const solve = timer.stop(); // timer.stop() returns the same value as the 'done' event listeners


```

## API

### Exported Types and Constants
Typescript typing files are provided with the library.


#### `TimerStatus`
These are the possible values that `timer.status` can have.
```typescript
enum TimerStatus = {
  Running = "RUNNING",
  Stopped = "STOPPED",
}
```

#### `SolvePhase`
```typescript
enum SolvePhase = {
  Inspecting = "INSPECTING",
  Solving = "SOLVING",
}
```

#### Event Types
A timer instance emits certain events and lets you subscribe to those event. Event handlers you register
with the timer will be called with the `event` objects described here.

```typescript
enum EventTypes = {
  InspectionWarning = "inspectionWarning",
  Penalty = "penalty",
  Tick = "tick",
  Done = "done",
}
```

##### `inspectionWarning` event and the `IInspectionWarningEvent` inteface
WCA rules state that during the 15 seconds of inspection time, the referee will first give a warning at 8 seconds
and another warning at 12 seconds. These warnings are expressed as objects implementing the `IInspectionWarningEvent`
interface.

```typescript
interface IInspectionWarningEvent {
  timeRemaining: number;
}
```

__Event properties__
* `timeRemaining` is the time remaining on the inspectiong timer in `milliseconds`

##### `penalty` event and the `IPenaltyEvent` interface
The total inspection time allowed by the WCA is `15 + 2` seconds. If the competitor hasn't started their solve by the
end of the initial 15 seconds, the solve is penalized by 2 seconds. If by the end of the total 17 seconds, the solve
still hasn't started, the solve is marked DNF.

(TODO: This looks clumsy, two fields that are mutually exclusive fields. Make them constants and just send the penalty
type with the event)

```typescript
enum PenaltyType {
  PlusTwo = "PENALTY_PLUS_TWO";
  DNF = "PENALTY_DNF";
};

interface IPenaltyEvent {
  type: PenaltyType
}
```
__Event properties__
* `type`: what are you supposed to write here if you have the interface and enum definition above?!


##### `tick` event and `ITickEvent` interface
```typescript
interface ITickEvent {
  phase: SolvePhase;
  time: number
}
```

__Event properties__
* `phase` is the current phase of the solve.
* `time` is the elapsed time since the last `status` change. Basically, this is the either the current inspection time
or the current solve time.

Note: If you have set `options.noInspect` to true, all `tick` events will have the phase `SolvePhase.Solving`.

#### `done` event and the `ISolveEndEvent` interface
(TODO: the interface name and the event name don't match. Then again I want this to be as thin a wrapper around `tiny-timer`
as possible)

This event will be emitted as a result of any of the following:
* calling `timer.stop()`
* calling `timer.abort()`
* not calling either of the above methods within `options.timeLimit` if `options.timeLimit` wasn't set to `0` during
initialization.

```typescript
interface ISolveEndEvent {
  inspectionTime: number;
  isDNF: boolean;
  penalized: boolean;
  solveTime: number;
}
```

__Event properties__
* `inspectionTime`: total inspection time for the solve in `milliseconds`.
* `isDNF`: `true` if a solve wasn't started before the total 17 seconds of inspection time or if `timer.abort()` was
called.
* `penalized`: will be `true` if this solve was penalized during the inspection phase.
* `solveTime`: the actual solve time in `milliseconds`


### Methods
#### constructor
```javascript
cubeTimer = new canCubeTimer({
  interval: 100,
  noInspect: false,
  timeLimit: 10 * 60,
  // event handlers
  onInspectionWarning: (timeRemaining) => {},
  onPenalty: (isDNF, timeRemaining) => {},
  onTick: (status, time, timeRemaining) => {},
  onDone: ({ inspectionTime, isDNF, penalized, solveTime }) => {},
})
```

Create a new timer instance. All options are optional. `can-cube-timer` instances are `EventEmitter`s.


__Arguments__
* `interval`: refresh interval in `milliseconds`.
* `noInspect`: disable inspection. Calling the `start` method will directly start the timer
  in with `status` set to `SOLVING`.
* `timeLimit`: the time limit for a solve in `seconds`. WCA default is 10 minutes. Set to `0` to disable

The remaining options `onInspectionWarning`, `onPenalty`, `onStateChange`, `onTick` and `onDone` are just another way of attaching
event listeners to the various events. 
Since a `CanCubeTimer` is an `EventEmitter`, so you can always add event listeners like:
```javascript
const timer = new canCubeTimer();
timer.on('inspectionWarning', () => {});
timer.on('penalty', () => {});
timer.on('tick', () => {});
timer.on('done', () => {});
```


#### start
Start the timer. This is meant to be a convenience method that you can call on a timer instance. Depending on the current
status of the timer, it will either start the inspection or the solve. 

If you try to call `start` on a timer that is already in the `SolvePhase.Solving` status an error will be thrown.
(TODO: Error definition)

_Disclaimer_
This is a stateful method that either calls the `startInspection` or the `startSolve` method. It is meant to make it possible
to write code that is less verbose.
It also lets you write code that is (in my opinion) fragile. It is better to use the more explicit methods.

If you try to call this method more than twice during a solve, an error will be thrown.
(TODO: Error defnition. Should be more specific than the previous one)

#### startInspection
Starts the inspection.

An error will be thrown
* if you try to call it more than once during a solve
* if you try to call it on a timer that was created with `options.noInspect` set to `true`
* if you try to call it on a timer that is in the `SolvePhase.Solving` phase


#### startSolve
Starts the actual solve. If you try to call it more than once an error will be thrown.

#### stop(): ISolveEndEvent
Stops the solve. An `EventTypes.Done` event will be emitted and the same event that is passed to registered event listeners
will be returned.

#### abort(): ISolveEndEvent
(TODO: Should there be a separate event for this?)
Aborts a solve. It stops the timer but the emitted `ISolveEndEvent` event object will have the `isDNF` field set to `true`



CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### 1.1.1 (2019-12-17)


src/lib/cube-timer.spec.ts
// tslint:disable:no-expression-statement

const fullTests = process.argv.slice(2)[0] === 'full';

function sleep(ms: number): Promise<number> {
  return new Promise(resolve => {
    setTimeout(resolve, ms);
  });
}

import test from 'ava';
import CanCubeTimer, { CanCubeError } from '../lib/cube-timer';
import { TimerStatus, EventType, TickEvent } from './types';

test('That a timer get initialized correctly without any parameters passed to the constructor', t => {
  const timer = new CanCubeTimer();
  t.is(timer.configuration.interval, 100);
  t.false(timer.configuration.noInspect);
  t.is(timer.configuration.timeLimit, 10 * 60);
});

test('Correctly initializes with the supplied options', t => {
  const timer = new CanCubeTimer({
    interval: 10,
    noInspect: true,
    timeLimit: 60,
  });

  t.is(timer.configuration.interval, 10);
  t.true(timer.configuration.noInspect);
  t.is(timer.configuration.timeLimit, 60);
});

test.cb('Starts inspection correctly and calls the supplied tick handler correctly', t => {
  const timer = new CanCubeTimer();
  timer.on(EventType.Tick, ({ status }: TickEvent) => {
    t.is(status, TimerStatus.Inspecting);
    t.end();
  });
  timer.startInspection();
});

test('Throws an error if you try to call startInspection on a timer initialized with options.noInspect set to true', t => {
  const timer = new CanCubeTimer({ noInspect: true });
  t.throws(
    timer.startInspection,
    CanCubeError,
    "Tried calling 'startInspection' with 'noInspect' option set to true!",
  );
  t.is(timer.status, TimerStatus.Stopped);
});

test('Throws an error if you try to call startInspection more than once during a solve', t => {
  const timer = new CanCubeTimer();
  timer.startInspection();

  t.throws(
    timer.startInspection,
    CanCubeError,
    "Tried calling 'startInspection' more than once during a single solve!",
  );
  t.is(timer.status, TimerStatus.Stopped);
});

test('Throws an error if you try to call startInspection while the timer status is TimerStatus.Solving', t => {
  // This should not be possible to achieve using the public API
  const timer = new CanCubeTimer();
  timer.startInspection();
  // TODO: Make this a real test once you have startSolve implemented
  t.is(true, true);
  //t.is(timer.status, TimerStatus.Stopped);
});

test('With inspection enabled, starts inspection, starts solve and reports correct data', t => {
  const timer = new CanCubeTimer();

  timer.startInspection();
  timer.startSolve();
  const solveData = timer.stop();

  t.not(solveData.inspectionTime, null);
  t.is(solveData.isDNF, false);
  t.is(solveData.aborted, false);
  t.is(solveData.penalized, false);
  t.not(solveData.solveTime, null);
});

test('Abort during inspection', async t => {
  const timer = new CanCubeTimer();
  timer.startInspection();
  await sleep(1000);

  const solveInformation = timer.abort();
  t.is(solveInformation.inspectionTime, null);
  t.true(solveInformation.isDNF);
  t.true(solveInformation.aborted);
  t.false(solveInformation.penalized);
  t.is(solveInformation.solveTime, null);
});

if (fullTests) {
  test('15+ inspection with +2 second penalty, solve normal', async t => {
    t.timeout(18000);
    const timer = new CanCubeTimer();
    timer.startInspection();
    await sleep(16500);
    timer.startSolve();
    const solveData = timer.stop();

    t.true(solveData.inspectionTime > 15000);
    t.false(solveData.isDNF);
    t.false(solveData.aborted);
    t.true(solveData.penalized);
    t.not(solveData.solveTime, null);
  });

  test.cb('DNF in inspection', t => {
    t.timeout(18500);
    const timer = new CanCubeTimer();
    timer.on(EventType.SolveEnd, solveInformation => {
      t.true(solveInformation.inspectionTime === 17000);
      t.true(solveInformation.isDNF);
      t.true(solveInformation.penalized);
      t.is(solveInformation.solveTime, null);
      t.is(timer.status, TimerStatus.Stopped);
      t.end();
    });
    timer.startInspection();
  });
}
test.todo('updates the timer status correctly and calls the registered listeners properly');

test.todo('.start() method behaves properly when inspection is enabled');

test.todo('.start() method behaves properly when inspection is disabled');

test.todo('aborting a solve during the solve returns correct data');


src/lib/types.ts
export enum TimerStatus {
  Stopped = 'STOPPED',
  Inspecting = 'INSPECTING',
  Solving = 'SOLVING',
}

export enum EventType {
  StatusChange = 'statusChange',
  InspectionWarning = 'inspectionWarning',
  Penalty = 'penalty',
  Tick = 'tick',
  SolveEnd = 'solveEnd',
}

export enum PenaltyType {
  PlusTwo = 'PENALTY_PLUS_TWO',
  DNF = 'PENALTY_DNF',
}

export interface PenaltyEvent {
  type: PenaltyType;
}

export interface InspectionWarningEvent {
  timeRemaining: number;
}

export interface TickEvent {
  status: TimerStatus;
  time: number;
}

export interface TimerConfiguration {
  interval: number;
  noInspect: boolean;
  timeLimit: number;
}

export interface SolveEndEvent {
  inspectionTime: number | null;
  isDNF: boolean;
  aborted: boolean | null;
  penalized: boolean;
  solveTime: number | null;
}

export interface Options {
  interval?: number;
  noInspect?: boolean;
  timeLimit?: number;
}

export enum TinyTimerStatus {
  Running = 'running',
  Paused = 'paused',
  Stopped = 'stopped',
}


src/lib/cube-timer.ts
import { EventEmitter } from 'events';
import TinyTimer from 'tiny-timer';
import {
  EventType,
  SolveEndEvent,
  Options,
  TinyTimerStatus,
  PenaltyType,
  TimerStatus,
  TimerConfiguration,
} from './types';

export class CanCubeError extends Error {
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

const defaultConfig = {
  interval: 100,
  noInspect: false,
  timeLimit: 10 * 60,
};
export default class CanCubeTimer extends EventEmitter {
  private inspectionTimer: TinyTimer | undefined;
  private solveTimer: TinyTimer;
  private solveInformation: SolveEndEvent | null = null;
  private inspectionTimerTickHandleCount = 0;
  private tickIntervalId?: NodeJS.Timer;

  public readonly configuration: TimerConfiguration;

  constructor(options: Options | undefined = {}) {
    super();

    // TODO: If I use the following commented line ava starts failing
    // this.configuration = Object.assign(defaultConfig, options);
    // The tests fail if they are ran in parallel but not in sequence.
    // That is interesting and I would like to know why
    this.configuration = {
      ...defaultConfig,
      ...options,
    };

    if (!this.configuration.noInspect) {
      this.inspectionTimer = new TinyTimer();
    }
    this.solveTimer = new TinyTimer({
      stopwatch: true,
      interval: this.configuration.timeLimit + 1000, // We don't need it to tick. TODO: Do we even need it?
    });

    this.initInternalTimers();
  }

  private startTicking = (): void => {
    if (!this.tickIntervalId) {
      this.tick();
      this.tickIntervalId = setInterval(this.tick, this.configuration.interval);
    }
  };

  private tick = (): void => {
    this.emit(EventType.Tick, {
      status: this.status,
      time:
        this.status === TimerStatus.Inspecting ? this.inspectionTimer.time : this.solveTimer.time,
    });
  };

  private emitInspectionWarning(): void {
    this.inspectionTimerTickHandleCount++;
    this.emit(EventType.InspectionWarning, {
      timeRemaining: this.inspectionTimer.time - 2000,
    });
  }

  private onInspectionTick = (time: number): void => {
    if (time < 10000 && this.inspectionTimerTickHandleCount === 0) {
      this.emitInspectionWarning();
    } else if (time < 5000 && this.inspectionTimerTickHandleCount === 1) {
      this.emitInspectionWarning();
    } else if (time < 2000 && this.inspectionTimerTickHandleCount === 2) {
      this.inspectionTimerTickHandleCount++;
      this.solveInformation.penalized = true;
      this.emit(EventType.Penalty, { type: PenaltyType.PlusTwo });
    }
  };

  private initInternalTimers = (): void => {
    if (!this.configuration.noInspect) {
      this.inspectionTimer.on('tick', this.onInspectionTick);

      this.inspectionTimer.on('done', () => {
        this.solveInformation.inspectionTime = 17000;
        this.stopWithDNF();
      });

      this.solveTimer.on('done', () => {
        this.solveInformation.solveTime = this.configuration.timeLimit;
      });
    }
  };

  private resetSolveInformation = (): void => {
    this.solveInformation = {
      inspectionTime: null,
      isDNF: false,
      aborted: false,
      penalized: false,
      solveTime: null,
    };
  };

  private stopInternalTimers = (): void => {
    !this.configuration.noInspect && this.inspectionTimer.stop();
    this.solveTimer.stop();
  };

  private resetTimer = (): void => {
    clearInterval(this.tickIntervalId);
    this.resetSolveInformation();
    this.stopInternalTimers();
    this.inspectionTimerTickHandleCount = 0;
  };

  private stopWithDNF = (): SolveEndEvent | null => {
    this.solveInformation.isDNF = true;
    return this.stop();
  };

  public abort = (): SolveEndEvent | null => {
    this.solveInformation.aborted = true;
    return this.stopWithDNF();
  };

  public stop = (): SolveEndEvent | null => {
    const solveEndInfo = Object.assign({}, this.solveInformation);
    if (this.status === TimerStatus.Solving) {
      solveEndInfo.solveTime = this.solveTimer.time;
    }

    this.emit(EventType.SolveEnd, solveEndInfo);
    this.emit(EventType.StatusChange, TimerStatus.Stopped);
    this.resetTimer();
    return solveEndInfo;
  };

  private throwError = (message: string): void => {
    this.resetTimer(); // Reset everything before we throw
    throw new CanCubeError(message);
  };

  public startInspection = (): void => {
    if (this.configuration.noInspect) {
      this.throwError("Tried calling 'startInspection' with 'noInspect' option set to true!");
    }

    if (this.inspectionTimer.status === TinyTimerStatus.Running) {
      this.throwError("Tried calling 'startInspection' more than once during a single solve!");
    }

    if (this.status === TimerStatus.Solving) {
      this.throwError("Tried calling 'startInspection' while in the 'Solving' phase!");
    }

    this.emit(EventType.StatusChange, TimerStatus.Inspecting);
    this.resetSolveInformation();

    this.inspectionTimer.start(17000); // This will run for at most 15 + 2;
    this.startTicking();
  };

  public startSolve = (): void => {
    if (this.solveTimer.status === TinyTimerStatus.Running) {
      this.throwError("Tried calling 'startSolve' while in the 'Solving' phase!");
    }

    if (!this.configuration.noInspect && this.status === TimerStatus.Inspecting) {
      this.solveInformation.inspectionTime = 17000 - this.inspectionTimer.time;
      this.inspectionTimer.stop();
      this.emit(EventType.StatusChange, TimerStatus.Solving);
    }

    this.solveTimer.start(this.configuration.timeLimit * 1000);
  };

  public start = (): void => {
    if (!this.configuration.noInspect && this.status === TimerStatus.Stopped) {
      this.startInspection();
    } else {
      this.startSolve();
    }
  };

  get status(): TimerStatus {
    if (!this.configuration.noInspect && this.inspectionTimer.status === TinyTimerStatus.Running) {
      return TimerStatus.Inspecting;
    }

    if (this.solveTimer.status === TinyTimerStatus.Running) {
      return TimerStatus.Solving;
    }

    return TimerStatus.Stopped;
  }

  get time(): number | null {
    if (this.status === TimerStatus.Inspecting) {
      return this.inspectionTimer.time;
    }

    if (this.status === TimerStatus.Solving) {
      return this.solveTimer.time;
    }

    return null;
  }
}


src/lib/number.ts
import { EventEmitter } from 'events';
import * as TinyTimer from 'tiny-timer';
/**
 * Multiplies a value by 2. (Also a full example of Typedoc's functionality.)
 *
 * ### Example (es module)
 * ```js
 * import { double } from 'typescript-starter'
 * console.log(double(4))
 * // => 8
 * ```
 *
 * ### Example (commonjs)
 * ```js
 * var double = require('typescript-starter').double;
 * console.log(double(4))
 * // => 8
 * ```
 *
 * @param value   Comment describing the `value` parameter.
 * @returns       Comment describing the return type.
 * @anotherNote   Some other value.
 */
export function double(value: number): number {
  return value * 2;
}

/**
 * Raise the value of the first parameter to the power of the second using the es7 `**` operator.
 *
 * ### Example (es module)
 * ```js
 * import { power } from 'typescript-starter'
 * console.log(power(2,3))
 * // => 8
 * ```
 *
 * ### Example (commonjs)
 * ```js
 * var power = require('typescript-starter').power;
 * console.log(power(2,3))
 * // => 8
 * ```
 */
export function power(base: number, exponent: number): number {
  // This is a proposed es7 operator, which should be transpiled by Typescript
  return base ** exponent;
}

enum TimerStatus {
  Running = 'RUNNING',
  Stopped = 'STOPPED'
}

enum SolvePhase {
  Inspecting = 'INSPECTING',
  Solving = 'SOLVING'
}

enum EventType {
  InspectionWarning = 'inspectionWarning',
  Penalty = 'penalty',
  Tick = 'tick',
  Done = 'done'
}

enum PenaltyType {
  PlusTwo = 'PENALTY_PLUS_TWO',
  DNF = 'PENALTY_DNF'
}

interface IPenaltyEvent {
  type: PenaltyType;
}

interface IInspectionWarningEvent {
  timeRemaining: number;
}

interface ITickEvent {
  phase: SolvePhase;
  timer: number;
}

interface ISolveEndEvent {
  inspectionTime?: number;
  isDNF: boolean;
  penalized: boolean;
  solveTime?: number;
}

interface Options {
  interval: number;
  noInspect: boolean;
  timeLimit: number;
}

enum TinyTimerStatus {
  Running = 'running',
  Paused = 'paused',
  Stopped = 'stopped'
}

export class CanCubeError extends Error {
  constructor(message?: string) {
    super(message);
    Object.setPrototypeOf(this, new.target.prototype);
  }
}

export class CanCubeTimer extends EventEmitter {
  private solvePhase: SolvePhase | null;
  private inspectionTimer: EventEmitter;
  private solveTimer: EventEmitter;
  private options: Options;

  private solveInformation: ISolveEndEvent | null;

  private inspectionTimerTickHandleCount = 0;

  constructor(
    options: Options = {
      interval: 100,
      noInspect: false,
      timeLimit: 10 * 60
    }
  ) {
    super();
    this.options = options;
  }

  private emitInspectionWarning() {
    this.inspectionTimerTickHandleCount++;
    this.emit(EventType.InspectionWarning, {
      timeRemaining: 15 - this.inspectionTimer.time
    });
  }

  private onInspectionTick = time => {
    if (time > 8000 && this.inspectionTimerTickHandleCount === 0) {
      this.emitInspectionWarning();
    } else if (time > 12000 && this.inspectionTimerTickHandleCount === 1) {
      this.emitInspectionWarning();
    } else if (time > 15000 && this.inspectionTimerTickHandleCount === 2) {
      // emit the +2 penalty
      this.emit(EventType.Penalty, { type: PenaltyType.PlusTwo });
    }
  };

  private initInternalTimers() {
    this.inspectionTimer = new TinyTimer();
    this.inspectionTimer.on('tick', this.onInspectionTick());

    this.inpectionTimer.on('done', () => {
      // emit DNF and solve end.
    });

    this.solveTimer = new TinyTimer({
      stopwatch: true,
      interval: this.options.timeLimit + 1000 // We don't need the tick event from tiny-timer at all.
    });

    this.solveTimer.on('done', () => {
      // emit DNF and solve end.
    });
  }

  private resetSolveInformation() {
    this.solveInformation = {
      inspectionTime: null,
      isDNF: false,
      penalized: false,
      solveTime: null
    };
  }

  public startInspection() {
    if (this.options.noInspect) {
      throw new CanCubeError(
        "Tried calling 'startInspection' with 'noInspect' option set to true!"
      );
    }

    if (this.inspectionTimer.status === TinyTimerStatus.Running) {
      throw new CanCubeError(
        "Tried calling 'startInspection' more than once during a single solve!"
      );
    }

    if (this.solvePhase === SolvePhase.Solving) {
      throw new CanCubeError(
        "Tried calling 'startInspection' while in the 'Solving' phase!"
      );
    }

    if (this.solveInformation === null) {
      this.resetSolveInformation();
    }

    this.inspectionTimer.start(17000); // This will run for at most 15 + 2;
  }

  public startSolve() {
    if (this.solveTimer.status === TinyTimerStatus.Running) {
      throw new CanCubeError(
        "Tried calling 'startSolve' while in the 'Solving' phase!"
      );
    }

    if (this.solvePhase === SolvePhase.Inspecting) {
      this.solveInformation.inspectionTime = this.inspectionTimer.time;
      this.inspectionTimer.stop();
    }

    this.solveTimer.start(this.options.timeLimit);
  }
}
